File: Week-02/Day-03/main.go
package main

import "fmt"

// Node represent a node in a doubly
type Node struct {
	data int
	prev *Node
	next *Node
}

// DoublyLinkedList represents a doubly linked list
type DoublyLinkedList struct {
	head *Node
	tail *Node
}

func main() {
	list := DoublyLinkedList{}
	list.InsertAtBeginnning(3)
	list.InsertAtBeginnning(71)
	list.InsertAtBeginnning(21)
	list.InsertAtEnd(69)

	//  prints all the nodes in the list from head to tail
	fmt.Println("List from head to tail: ")
	list.PrintListForward()

	//  prints all the nodes in the list from tail to head
	fmt.Println("List from tail to head: ")
	list.PrintListBackward()

	list.DeleteNode(21)
	fmt.Println("Liste after deleting node with data 21: ")
	list.PrintListForward()
}

// InsertAtBeginning inserts a new node at the beginning of the list
func (list *DoublyLinkedList) InsertAtBeginnning(data int) {
	newNode := &Node{data: data}
	if list.head == nil {
		list.head = newNode
		list.tail = newNode
	} else {
		newNode.next = list.head
		list.head.prev = newNode
		list.head = newNode
	}
}

// InsertAtEnd inserts a new node at the end of the list
func (list *DoublyLinkedList) InsertAtEnd(data int) {
	newNode := &Node{data: data}
	if list.tail == nil {
		list.head = newNode
		list.tail = newNode
	} else {
		list.tail.next = newNode
		newNode.prev = list.tail
		list.tail = newNode
	}
}

// DeleteNode deletes a node with specific data from the list
func (list *DoublyLinkedList) DeleteNode(data int) {
	temp := list.head
	for temp != nil && temp.data != data {
		temp = temp.next
	}
	if temp == nil {
		fmt.Println("Node not found")
		return
	}

	if temp.prev != nil {
		temp.prev.next = temp.next
	} else {
		list.head = temp.next
	}
	if temp.next != nil {
		temp.next.prev = temp.prev
	} else {
		list.tail = temp.prev
	}
}

// PrintListForward prints all the nodes in the list from head to tail
func (list *DoublyLinkedList) PrintListForward() {
	temp := list.head
	for temp != nil {
		fmt.Printf("%d -> ", temp.data)
		temp = temp.next
	}
	fmt.Println("nil")
}

// PrintListBackward prints all the nodes in the list from tail to head
func (list *DoublyLinkedList) PrintListBackward() {
	temp := list.tail
	for temp != nil {
		fmt.Printf("%d -> ", temp.data)
		temp = temp.prev
	}
	fmt.Println("nil")
}

File: Week-02/Day-03/main.cpp
#include <iostream>
using namespace std;

class Node {
public:
  int data;
  Node* next;

  Node(int val) : data(val), next(nullptr) {}
};

// SinglyLinkedList class definition
class SinglyLinkedList {
private:
  Node* head;

public:
  SinglyLinkedList() : head(nullptr) {}

  // insertAtBeginning
  void insertAtBeginning(int data) {
    Node* newNode = new Node(data);
    newNode->next = head;
    head = newNode;
  }

// insertAtEnd
  void insertAtEnd(int val){
    Node* newNode = new Node(val);
    if(!head) {
      head = newNode;
      return;
    }
    Node* temp = head;
    while(temp->next) {
      temp = temp->next;
    }
    temp->next = newNode;
  }

  // detectingLoop 
  bool detectLoop(){
    Node* slow = head;
    Node* fast = head;
  
    while (slow && fast && fast->next) {
      slow = slow->next;
      fast = fast->next->next;
      if(slow == fast) {
        return true;
      }
    }
    return false;
  }

  // remove a loop
  void removeloop(Node* loopNode, Node* head){
    Node* ptr1 = head;
    Node* ptr2;
    while (1) {
      ptr2 = loopNode;
      while (ptr2->next == ptr1) {
        break;
      }
      ptr1 = ptr1->next;
    }
    ptr2->next = nullptr; // break the loop
  }

  // Find the middle
  Node* findMiddle(Node* head){
    if(!head) return nullptr;
    Node* slow = head;
    Node* fast = head;
    while(fast && fast->next) {
      slow = slow->next;
      fast = fast->next->next;
    }
    return slow; // this is middle node 
  }

  // display linked list
  void printList(){
    Node* temp = head;
    while(temp->next){
      cout << temp->data << " ";
      temp = temp->next;
    }
    cout << "nil" << endl;
  }

};


int main(){
  SinglyLinkedList list;

  // inserting element at beginning
  list.insertAtBeginning(21);
  list.insertAtBeginning(71);
  list.insertAtBeginning(7);
  list.insertAtBeginning(33);
  list.insertAtBeginning(21);

  // inserting element at end
  list.insertAtEnd(9);

  // print the list
  list.printList();

  // Loop checking
  if (list.detectLoop()) 
    cout <<  "Loop detected\n";
  else 
    cout << "No loop\n";

}

